File: BladiOS/Library.swift
import SwiftUI

/// Blad's files on iPhone and iPad. A space is a folder picked in the Files app and remembered
/// with a security-scoped bookmark, so notes can live in iCloud Drive next to the Mac's.
@Observable
final class Library {
    static let shared = Library()

    struct Space: Identifiable, Hashable {
        let url: URL
        var id: URL { url }
        var name: String { url.lastPathComponent }
    }

    /// A page moved to a new name; the open screen follows it.
    struct Rename: Equatable {
        let from: URL
        let to: URL
    }

    private(set) var spaces: [Space] = []
    private(set) var trees: [URL: [FileNode]] = [:]
    private(set) var lastRename: Rename?
    var errorMessage: String?

    @ObservationIgnored private var documents: [URL: DocumentModel] = [:]
    @ObservationIgnored private let defaults = UserDefaults.standard
    private let bookmarksKey = "spaceBookmarks"

    private init() {
        restore()
    }

    // MARK: - Spaces

    func addSpace(_ url: URL) {
        guard url.startAccessingSecurityScopedResource() else {
            errorMessage = "Blad kreeg geen toegang tot \(url.lastPathComponent)."
            return
        }
        let url = url.standardizedFileURL
        guard !spaces.contains(where: { $0.url == url }) else { return }
        spaces.append(Space(url: url))
        persist()
        refresh(url)
    }

    func removeSpace(_ space: Space) {
        spaces.removeAll { $0 == space }
        trees[space.url] = nil
        space.url.stopAccessingSecurityScopedResource()
        persist()
    }

    func space(containing url: URL) -> Space? {
        spaces.first { url.path == $0.url.path || url.path.hasPrefix($0.url.path + "/") }
    }

    func refreshAll() {
        for space in spaces { refresh(space.url) }
        for document in documents.values { document.reloadFromDisk() }
    }

    func refresh(containing url: URL) {
        if let space = space(containing: url) { refresh(space.url) }
    }

    private func refresh(_ folder: URL) {
        let sort = FileSort(rawValue: defaults.string(forKey: Pref.sortOrder) ?? "") ?? .name
        Task {
            let nodes = await Task.detached(priority: .userInitiated) { () -> [FileNode] in
                Library.downloadPlaceholders(in: folder)
                return FileNode.scan(folder, sort: sort)
            }.value
            if trees[folder] != nodes { trees[folder] = nodes }
        }
    }

    /// iCloud Drive keeps files that aren't on this device yet as hidden `.name.icloud` placeholders.
    nonisolated private static func downloadPlaceholders(in folder: URL) {
        guard let enumerator = FileManager.default.enumerator(at: folder, includingPropertiesForKeys: nil) else { return }
        while let url = enumerator.nextObject() as? URL {
            let name = url.lastPathComponent
            guard name.hasPrefix("."), name.hasSuffix(".icloud") else { continue }
            let original = url.deletingLastPathComponent().appendingPathComponent(String(name.dropFirst().dropLast(7)))
            try? FileManager.default.startDownloadingUbiquitousItem(at: original)
        }
    }

    // MARK: - Pages

    func document(for url: URL) -> DocumentModel? {
        if let document = documents[url] { return document }
        guard let document = try? DocumentModel(url: url) else { return nil }
        documents[url] = document
        return document
    }

    func saveAll() {
        documents.values.forEach { $0.save() }
    }

    /// A new, empty page. It takes its name from its first heading when you leave it.
    func newPage(in folder: URL? = nil) -> URL? {
        guard let folder = folder ?? spaces.first?.url else { return nil }
        let url = uniqueURL(in: folder, base: "Naamloos", pathExtension: "md")
        do {
            try Data().write(to: url)
        } catch {
            present(error, "Kon geen nieuwe pagina maken")
            return nil
        }
        refresh(containing: url)
        document(for: url)?.namesItselfFromHeading = true
        return url
    }

    /// Renames a new page after its first heading. Done when leaving the page, not while typing,
    /// so the keyboard never disappears mid-sentence.
    func finishEditing(_ document: DocumentModel) {
        document.save()
        guard document.namesItselfFromHeading, let heading = firstHeading(in: document.text) else { return }
        let name = sanitized(heading)
        guard !name.isEmpty, name != document.title else { return }
        let destination = document.url.deletingLastPathComponent().appendingPathComponent(name).appendingPathExtension("md")
        guard !FileManager.default.fileExists(atPath: destination.path) else { return }
        move(document.url, to: destination, keepsAutoNaming: true)
    }

    @discardableResult
    func rename(_ url: URL, to name: String) -> URL? {
        let clean = sanitized(name)
        guard !clean.isEmpty else { return nil }
        var destination = url.deletingLastPathComponent().appendingPathComponent(clean)
        let isFolder = (try? url.resourceValues(forKeys: [.isDirectoryKey]))?.isDirectory == true
        if !isFolder {
            destination.appendPathExtension(url.pathExtension.isEmpty ? "md" : url.pathExtension)
        }
        guard destination != url else { return url }
        return move(url, to: destination, keepsAutoNaming: false) ? destination : nil
    }

    func delete(_ url: URL) {
        documents = documents.filter { $0.key != url && !$0.key.path.hasPrefix(url.path + "/") }
        do {
            do {
                try FileManager.default.trashItem(at: url, resultingItemURL: nil)
            } catch {
                // Not every location has a trash, "On My iPhone" for one.
                try FileManager.default.removeItem(at: url)
            }
        } catch {
            present(error, "Kon \(url.lastPathComponent) niet verwijderen")
            return
        }
        refresh(containing: url)
    }

    @discardableResult
    private func move(_ source: URL, to destination: URL, keepsAutoNaming: Bool) -> Bool {
        do {
            try FileManager.default.moveItem(at: source, to: destination)
        } catch {
            present(error, "Kon de naam niet wijzigen")
            return false
        }
        for (url, document) in documents where url == source || url.path.hasPrefix(source.path + "/") {
            let moved = URL(fileURLWithPath: destination.path + url.path.dropFirst(source.path.count))
            document.move(to: moved)
            if url == source, !keepsAutoNaming { document.namesItselfFromHeading = false }
            documents[url] = nil
            documents[moved] = document
        }
        lastRename = Rename(from: source, to: destination)
        refresh(containing: destination)
        return true
    }

    // MARK: - Links and search

    /// Pages offered in the `[[` link picker, most recently changed first.
    var pageRefs: [PageRef] {
        var pages: [(ref: PageRef, modified: Date)] = []
        func walk(_ nodes: [FileNode], location: String) {
            for node in nodes {
                if node.isDirectory {
                    walk(node.children ?? [], location: location + " › " + node.name)
                } else {
                    pages.append((PageRef(title: node.name, location: location), node.modified ?? .distantPast))
                }
            }
        }
        for space in spaces {
            walk(trees[space.url] ?? [], location: space.name)
        }
        return pages.sorted { $0.modified > $1.modified }.map(\.ref)
    }

    /// The page a `[[link]]` points to, preferring the linking page's own space.
    /// A link to a page that doesn't exist yet creates it next to the linking page.
    func openPage(named name: String, from source: URL) -> URL? {
        let target = (name.trimmingCharacters(in: .whitespaces) as NSString).lastPathComponent
        guard !target.isEmpty else { return nil }

        func pages(in nodes: [FileNode]) -> [URL] {
            nodes.flatMap { $0.isDirectory ? pages(in: $0.children ?? []) : [$0.url] }
        }
        let matches = spaces.flatMap { pages(in: trees[$0.url] ?? []) }.filter {
            $0.deletingPathExtension().lastPathComponent.compare(target, options: [.caseInsensitive, .diacriticInsensitive]) == .orderedSame
        }
        let sourceSpace = space(containing: source)
        if let url = matches.first(where: { space(containing: $0) == sourceSpace }) ?? matches.first {
            return url
        }

        let url = source.deletingLastPathComponent().appendingPathComponent(sanitized(target)).appendingPathExtension("md")
        do {
            try "# \(target)\n\n".write(to: url, atomically: true, encoding: .utf8)
        } catch {
            present(error, "Kon de pagina \(target) niet maken")
            return nil
        }
        refresh(containing: url)
        return url
    }

    func buildSearchIndex() async -> [SearchEntry] {
        let folders = spaces.map(\.url)
        let openTexts = Dictionary(documents.map { ($0.key, $0.value.text) }, uniquingKeysWith: { first, _ in first })
        return await Task.detached(priority: .userInitiated) {
            SearchIndex.build(workspaces: folders, looseFiles: [], openTexts: openTexts)
        }.value
    }

    func backlinks(to document: DocumentModel) async -> [Backlink] {
        let entries = await buildSearchIndex()
        let title = document.title
        let url = document.url
        return await Task.detached(priority: .utility) {
            WikiLinks.backlinks(to: title, excluding: url, in: entries)
        }.value
    }

    // MARK: - Helpers

    private func persist() {
        defaults.set(spaces.compactMap { try? $0.url.bookmarkData() }, forKey: bookmarksKey)
    }

    private func restore() {
        let bookmarks = defaults.array(forKey: bookmarksKey) as? [Data] ?? []
        for bookmark in bookmarks {
            var isStale = false
            guard let url = try? URL(resolvingBookmarkData: bookmark, bookmarkDataIsStale: &isStale),
                  url.startAccessingSecurityScopedResource() else { continue }
            spaces.append(Space(url: url.standardizedFileURL))
        }
        // Writing the bookmarks again replaces any that went stale.
        persist()
        refreshAll()
    }

    private func firstHeading(in text: String) -> String? {
        let firstLine = text.split(whereSeparator: \.isNewline).first { !$0.allSatisfy(\.isWhitespace) }
        guard let line = firstLine?.trimmingCharacters(in: .whitespaces), line.hasPrefix("#") else { return nil }
        return line.drop { $0 == "#" }.trimmingCharacters(in: .whitespaces)
    }

    private func sanitized(_ name: String) -> String {
        let cleaned = name
            .replacingOccurrences(of: "/", with: "-")
            .replacingOccurrences(of: ":", with: "-")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return String(cleaned.prefix(80))
    }

    private func uniqueURL(in folder: URL, base: String, pathExtension: String) -> URL {
        var number = 1
        while true {
            let url = folder
                .appendingPathComponent(number == 1 ? base : "\(base) \(number)")
                .appendingPathExtension(pathExtension)
            if !FileManager.default.fileExists(atPath: url.path) { return url.standardizedFileURL }
            number += 1
        }
    }

    private func present(_ error: Error, _ message: String) {
        errorMessage = "\(message).\n\n\(error.localizedDescription)"
    }
}

File: windows/Blad.App/Library.cs
using Blad.Core.Markdown;
using Blad.Core.Pages;
using Blad.Core.Search;
using Microsoft.UI.Dispatching;
using Windows.Storage;

namespace Blad;

/// <summary>A page as the link picker and search show it.</summary>
public sealed record PageInfo(string Path, string Title, string Location, DateTime Modified);

/// <summary>
/// Spaces, their pages and the open documents. A space is a plain folder, one per project or course,
/// so pages stay .md files that also open on the Mac, the iPhone, in git or in any editor.
/// </summary>
public sealed class Library
{
    public static Library Shared { get; } = new();

    private readonly Dictionary<string, PageDocument> documents = new(StringComparer.OrdinalIgnoreCase);
    private readonly Dictionary<string, IReadOnlyList<PageNode>> trees = new(StringComparer.OrdinalIgnoreCase);
    private readonly HashSet<PageDocument> unsaved = [];
    private DispatcherQueueTimer? saveTimer;

    /// <summary>The page tree changed; the sidebar rebuilds.</summary>
    public event Action? TreesChanged;

    /// <summary>A page or folder moved from the first path to the second; open tabs follow it.</summary>
    public event Action<string, string>? PageMoved;

    public event Action<string>? ErrorOccurred;

    public IReadOnlyList<string> Spaces => AppSettings.Current.Spaces;

    public IReadOnlyList<PageNode> Tree(string space) => trees.TryGetValue(space, out var nodes) ? nodes : [];

    private static PageSort Sort => AppSettings.Current.Sort == "modified" ? PageSort.Modified : PageSort.Name;

    // MARK: Spaces

    public async Task RefreshAsync()
    {
        var sort = Sort;
        var spaces = Spaces.ToList();
        var scanned = await Task.Run(() => spaces.Select(space => (space, PageFiles.Scan(space, sort))).ToList());
        trees.Clear();
        foreach (var (space, nodes) in scanned) trees[space] = nodes;
        foreach (var document in documents.Values) document.ReloadFromDisk();
        TreesChanged?.Invoke();
    }

    public void AddSpace(string folder)
    {
        if (Spaces.Any(space => SamePath(space, folder))) return;
        AppSettings.Current.Spaces.Add(folder);
        AppSettings.Current.Save();
        _ = RefreshAsync();
    }

    /// <summary>Makes a folder for a new project or course inside <paramref name="parent"/> and opens it as a space.</summary>
    public string? CreateSpace(string name, string parent)
    {
        var clean = PageFiles.SafeName(name);
        if (clean.Length == 0) return null;
        try
        {
            var folder = PageFiles.UniquePath(parent, clean, null);
            Directory.CreateDirectory(folder);
            AddSpace(folder);
            return folder;
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            Report($"Kon de ruimte {clean} niet maken.", exception);
            return null;
        }
    }

    public void RemoveSpace(string space)
    {
        AppSettings.Current.Spaces.RemoveAll(item => SamePath(item, space));
        AppSettings.Current.Save();
        trees.Remove(space);
        TreesChanged?.Invoke();
    }

    public string? SpaceContaining(string path) => Spaces.FirstOrDefault(space => IsInside(path, space));

    // MARK: Pages

    public PageDocument? Document(string path)
    {
        if (documents.TryGetValue(path, out var open)) return open;
        try
        {
            var document = PageDocument.Open(path);
            documents[path] = document;
            return document;
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            Report($"Kon {Path.GetFileName(path)} niet openen.", exception);
            return null;
        }
    }

    /// <summary>Saves the page shortly after typing stops.</summary>
    public void ScheduleSave(PageDocument document)
    {
        unsaved.Add(document);
        if (saveTimer is null)
        {
            saveTimer = DispatcherQueue.GetForCurrentThread().CreateTimer();
            saveTimer.Interval = TimeSpan.FromMilliseconds(700);
            saveTimer.IsRepeating = false;
            saveTimer.Tick += (_, _) => SaveAll();
        }
        saveTimer.Stop();
        saveTimer.Start();
    }

    public void SaveAll()
    {
        foreach (var document in unsaved.ToList())
        {
            try
            {
                document.Save();
                unsaved.Remove(document);
            }
            catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
            {
                Report($"Kon {document.Title} niet bewaren.", exception);
            }
        }
    }

    /// <summary>A new, empty page. It takes its name from its first heading when you leave it.</summary>
    public string? NewPage(string? folder)
    {
        folder ??= Spaces.FirstOrDefault();
        if (folder is null) return null;
        try
        {
            var path = PageFiles.UniquePath(folder, "Naamloos", ".md");
            File.WriteAllText(path, "");
            if (Document(path) is { } document) document.NamesItselfFromHeading = true;
            _ = RefreshAsync();
            return path;
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            Report("Kon geen nieuwe pagina maken.", exception);
            return null;
        }
    }

    public string? NewFolder(string parent)
    {
        try
        {
            var path = PageFiles.UniquePath(parent, "Nieuwe map", null);
            Directory.CreateDirectory(path);
            _ = RefreshAsync();
            return path;
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            Report("Kon geen map maken.", exception);
            return null;
        }
    }

    /// <summary>Renames a new page after its first heading. Done when leaving the page, so typing is never interrupted.</summary>
    public void FinishEditing(PageDocument document)
    {
        SaveAll();
        if (!document.NamesItselfFromHeading || document.FirstHeading() is not { } heading) return;
        var name = PageFiles.SafeName(heading);
        if (name.Length == 0 || name == document.Title) return;
        var destination = Path.Combine(Path.GetDirectoryName(document.Path)!, name + ".md");
        if (!File.Exists(destination)) Move(document.Path, destination, keepsAutoNaming: true);
    }

    public string? Rename(string path, string name)
    {
        var clean = PageFiles.SafeName(name);
        if (clean.Length == 0) return null;
        var isFolder = Directory.Exists(path);
        var destination = Path.Combine(Path.GetDirectoryName(path)!, clean + (isFolder ? "" : Path.GetExtension(path)));
        return SamePath(destination, path) ? path : Move(path, destination, keepsAutoNaming: false);
    }

    /// <summary>Moves a page or folder to the Recycle Bin, so it can be put back.</summary>
    public async Task DeleteAsync(string path)
    {
        foreach (var key in documents.Keys.Where(key => IsInside(key, path)).ToList())
        {
            unsaved.Remove(documents[key]);
            documents.Remove(key);
        }
        try
        {
            IStorageItem item = Directory.Exists(path)
                ? await StorageFolder.GetFolderFromPathAsync(path)
                : await StorageFile.GetFileFromPathAsync(path);
            await item.DeleteAsync(StorageDeleteOption.Default);
        }
        catch (Exception exception)
        {
            Report($"Kon {Path.GetFileName(path)} niet verwijderen.", exception);
        }
        await RefreshAsync();
    }

    private string? Move(string source, string destination, bool keepsAutoNaming)
    {
        try
        {
            if (Directory.Exists(source)) Directory.Move(source, destination);
            else File.Move(source, destination);
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            Report("Kon de naam niet wijzigen.", exception);
            return null;
        }

        foreach (var (path, document) in documents.Where(entry => IsInside(entry.Key, source)).ToList())
        {
            var moved = destination + path[source.Length..];
            document.MoveTo(moved);
            if (SamePath(path, source) && !keepsAutoNaming) document.NamesItselfFromHeading = false;
            documents.Remove(path);
            documents[moved] = document;
        }
        PageMoved?.Invoke(source, destination);
        _ = RefreshAsync();
        return destination;
    }

    // MARK: Links and search

    /// <summary>Every page in the spaces, most recently changed first: what the link picker offers.</summary>
    public IReadOnlyList<PageInfo> Pages()
    {
        var pages = new List<PageInfo>();
        void Walk(IEnumerable<PageNode> nodes, string location)
        {
            foreach (var node in nodes)
            {
                if (node.IsFolder) Walk(node.Children, location + " › " + node.Name);
                else pages.Add(new PageInfo(node.Path, node.Name, location, node.Modified));
            }
        }
        foreach (var space in Spaces) Walk(Tree(space), Path.GetFileName(space));
        return pages.OrderByDescending(page => page.Modified).ToList();
    }

    /// <summary>The page a [[link]] points to, preferring the linking page's own space. A link to a page that
    /// doesn't exist yet creates it next to the linking page.</summary>
    public string? OpenPage(string name, string fromPath)
    {
        var target = Path.GetFileName(name.Trim());
        if (target.Length == 0) return null;

        var matches = Pages().Where(page => WikiLinks.SameName(page.Title, target)).ToList();
        var fromSpace = SpaceContaining(fromPath);
        var match = matches.FirstOrDefault(page => SpaceContaining(page.Path) == fromSpace) ?? matches.FirstOrDefault();
        if (match is not null) return match.Path;

        try
        {
            var path = Path.Combine(Path.GetDirectoryName(fromPath)!, PageFiles.SafeName(target) + ".md");
            File.WriteAllText(path, $"# {target}\n\n");
            _ = RefreshAsync();
            return path;
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            Report($"Kon de pagina {target} niet maken.", exception);
            return null;
        }
    }

    public Task<List<SearchEntry>> BuildSearchIndexAsync()
    {
        var spaces = Spaces.ToList();
        var openTexts = documents.ToDictionary(entry => entry.Key, entry => entry.Value.Text, StringComparer.OrdinalIgnoreCase);
        return Task.Run(() => SearchIndex.Build(spaces, openTexts));
    }

    public async Task<IReadOnlyList<Backlink>> BacklinksAsync(PageDocument document)
    {
        var entries = await BuildSearchIndexAsync();
        var (title, path) = (document.Title, document.Path);
        return await Task.Run(() => WikiLinks.Backlinks(title, path, entries));
    }

    // MARK: Helpers

    public static bool SamePath(string a, string b) =>
        string.Equals(Path.TrimEndingDirectorySeparator(a), Path.TrimEndingDirectorySeparator(b), StringComparison.OrdinalIgnoreCase);

    public static bool IsInside(string path, string folder) =>
        SamePath(path, folder) || path.StartsWith(Path.TrimEndingDirectorySeparator(folder) + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase);

    private void Report(string message, Exception exception) => ErrorOccurred?.Invoke($"{message}\n\n{exception.Message}");
}
